json/lex_misc.mbt

```diff
@@ -13,37 +13,53 @@
 // limitations under the License.
 
 ///|
-fn read_char(ctx : ParseContext) -> Char? {
-  if ctx.offset < ctx.end_offset {
-    let c = ctx.input[ctx.offset]
-    ctx.offset += 1
-    let c1 = c.to_int()
-    if c1 >= 0xD800 && c1 <= 0xDBFF {
-      if ctx.offset < ctx.end_offset {
-        let c2 = ctx.input[ctx.offset].to_int()
-        if c2 >= 0xDC00 && c2 <= 0xDFFF {
-          ctx.offset += 1
-          let c3 = (c1 << 10) + c2 - 0x35fdc00
-          return Some(Char::from_int(c3))
-        }
+fn peek(self : ParseContext) -> Char? {
+  match self.view {
+    [] => None
+    [c, ..] => Some(c)
+  }
+}
+
+///|
+fn bump(self : ParseContext, builder? : Builder) -> Unit {
+  match self.view {
+    [] =>
+      // this should be a misuse of the peek & bump api
+      abort("empty view")
+    [c, .. rest] => {
+      if c == '\n' {
+        self.line += 1
```

Comment:
The new version added overhead at every bump(s). Previous version is highly optimized and considered reasonable. In most cases, the inputs of `json.parse` are well-formed json. So we don't need report error with locations.  

Do we have some benchmark? I think the refactoring of one of motivations "avoid substring/write_substring" should not mixed with refactoring of location tracking. It's better to make these change into separated commits. So we can measure them separately.